internal/appconfig/watcher.go

```diff
@@ -662,14 +687,32 @@ func buildAppConfig(
 		}
 	}
 
+	var ectx hcl.EvalContext
+
 	// If we have no dynamic values, then we just return the static ones.
 	if len(dynamic) == 0 {
-		return static
+		return expandStaticVars(log, &ectx, staticVars)
 	}
 
+	// The way this next bit works is that any static values that referenced
+	// other static values have already been expanded before they make it this far,
+	// which means that if a static variable still contains an HCL template, it's
+	// going to reference a dynamic variable. And because dynamic variables can't
+	// reference other variables, the job is pretty easy.
```

Comment:
Took me a sec to consider this but you're of course absolutely right... since dynamic vars can't reference anything by their nature (just fundamentally), this simplifies things a LOT. This is a really nice constraint that I didn't consider before! 👍

Reply:
I realized it as I started to code it up and did a little fist pump in the air.